private/bufpkg/bufplugin/bufplugindocker/docker.go

```diff
@@ -0,0 +1,191 @@
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+type Client interface {
+	// Build creates a Docker image for the plugin using the Dockerfile.plugin and plugin config.
+	Build(ctx context.Context, dockerfile io.Reader, config *bufpluginconfig.Config, params BuildParams) (*BuildResponse, error)
+	// Push the Docker image to the remote registry.
+	Push(ctx context.Context, image string, auth RegistryAuth) (*PushResponse, error)
+	// Delete removes the Docker image from local Docker Engine.
+	Delete(ctx context.Context, image string) (*DeleteResponse, error)
+	// Close releases any resources used by the underlying Docker client.
+	Close() error
+}
+
+type BuildParams struct {
+	// ConfigDirPath specifies the location where the .buildkit_node_id should be persisted.
+	// This is used to establish a session when building images using buildkit.
+	// If empty, the node id will be generated randomly and not persisted to disk.
+	ConfigDirPath string
+}
+
+type BuildResponse struct {
+	// Image contains the Docker image name in the local Docker engine including the tag (i.e. plugins.buf.build/library/some-plugin:<id>, where <id> is a random id).
+	// It is created from the bufpluginconfig.Config's Name.IdentityString() and a unique id.
+	Image string
+	// Digest specifies the Docker image digest in the format <hash_algorithm>:<hash>.
+	// Example: sha256:65001659f150f085e0b37b697a465a95cbfd885d9315b61960883b9ac588744e
+	Digest string
+}
+
+type PushResponse struct{}
+
+type DeleteResponse struct{}
+
+type dockerAPIClient struct {
+	cli    *client.Client
+	logger *zap.Logger
+}
+
+var _ Client = (*dockerAPIClient)(nil)
+
+func (d *dockerAPIClient) Build(ctx context.Context, dockerfile io.Reader, pluginConfig *bufpluginconfig.Config, params BuildParams) (*BuildResponse, error) {
+	buildkitSession, err := createSession(fmt.Sprintf("%s/%s", pluginConfig.Name.Owner(), pluginConfig.Name.Plugin()), params.ConfigDirPath, zap.L())
+	if err != nil {
+		return nil, err
+	}
+
+	dockerContext, err := createDockerContext(dockerfile)
+	if err != nil {
+		return nil, err
+	}
+
+	eg, errGroupCtx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		return buildkitSession.Run(errGroupCtx, func(ctx context.Context, proto string, meta map[string][]string) (net.Conn, error) {
+			return d.cli.DialHijack(ctx, "/session", proto, meta)
+		})
+	})
+
+	buildID := stringid.GenerateRandomID()
+	imageName := pluginConfig.Name.IdentityString() + ":" + buildID
+	eg.Go(func() error {
+		defer buildkitSession.Close()
+
+		response, err := d.cli.ImageBuild(ctx, dockerContext, types.ImageBuildOptions{
+			Tags:       []string{imageName},
+			Dockerfile: "Dockerfile",
+			Platform:   "linux/amd64",
+			Labels: map[string]string{
+				"build.buf.plugins.config.remote": pluginConfig.Name.Remote(),
+				"build.buf.plugins.config.owner":  pluginConfig.Name.Owner(),
+				"build.buf.plugins.config.name":   pluginConfig.Name.Plugin(),
+			},
```

Comment:
These were default labels that I though might make it easier on the BSR side of things to determine more information about published images (since they'll be published with a unique tag).

Reply:
What happens if the same image digest is actually produced by multiple plugins? Does it actually make sense to assign this to a single remote/owner/name combination? For example, the `buf.build/alex/go` and `buf.build/philip/go` plugins might use the same `Dockerfile.plugin`.

Or are you saying that we'll always only have a unique representation here via the unique tag, regardless of the image digest?

Reply:
We might need to be mindful of the labels that we define here - if it can ever be potentially part of our API, we need to make sure not to break them. That could put a little more weight on this decision, but that might not actually be a real concern.